server/item.go

```diff
@@ -76,84 +78,249 @@ func (s *RESTServer) SlotHandler(w http.ResponseWriter, r *http.Request, ps http
 	s.getblob(w, r, id, items.BlobID(bid))
 }
 
+// getblob will find the given blob, either in the cache or on
+// tape, and then send it as a response. If there is an error, it
+// will return an error response.
 func (s *RESTServer) getblob(w http.ResponseWriter, r *http.Request, id string, bid items.BlobID) {
 	key := fmt.Sprintf("%s+%04d", id, bid)
-	cacheContents, length, err := s.Cache.Get(key)
-	if err != nil {
+	firsttime := true
+retry:
+	content, err := s.findContent(key, id, bid, r.Method == "GET")
+	if err == items.ErrNoStore {
+		w.WriteHeader(503)
+		fmt.Fprintln(w, err)
+		return
+	} else if err == items.ErrDeleted {
+		w.WriteHeader(404)
+		fmt.Fprintln(w, err)
+		return
+	} else if _, ok := err.(items.NoBlobError); ok {
+		w.WriteHeader(404)
+		fmt.Fprintln(w, err)
+		return
+	} else if err != nil {
+		// figure out which errors are really 404 errors
+		log.Println("getblob", key, err)
 		w.WriteHeader(500)
 		fmt.Fprintln(w, err)
 		return
 	}
-	var src io.Reader
-	if cacheContents != nil {
-		nCacheHit.Add(1)
-		log.Printf("Cache Hit %s", key)
-		w.Header().Set("X-Cached", "1")
-		defer cacheContents.Close()
-		// need to wrap this since Cache.Get returns a ReadAtCloser
-		src = store.NewReader(cacheContents)
-	} else {
-		// cache miss...load from main store, AND put into cache
-		nCacheMiss.Add(1)
-		log.Printf("Cache Miss %s", key)
-
-		// If Item Store Use disabled, and not in cache, return 503
-		if !s.useTape {
-			w.WriteHeader(503)
-			fmt.Fprintln(w, items.ErrNoStore)
-			return
+	switch content.status {
+	case ContentCached:
+		if firsttime {
+			nCacheHit.Add(1)
+			log.Println("Cache Hit", key)
 		}
-
-		blobinfo, err := s.Items.BlobInfo(id, bid)
-		if err != nil {
-			w.WriteHeader(404)
-			fmt.Fprintln(w, err)
+		w.Header().Set("X-Cached", "1")
```

Comment:
Are the values of "X-Cached" an enumerable? Can these be constants instead? It looks like `"1"` and `"2"` are unspecified magic values.

Reply:
They are documented here: https://github.com/ndlib/bendo/blob/master/architecture/api.md#see-if-a-file-is-in-the-cache

Reply:
I understand they are documented, but in code magic strings should be constantized to convey their intention.

Reply:
Since they are only used in one place, it seems like that would make the code less readable by adding another level of indirection. (I totally see the point if the constants are used in many places, or for loop boundries).

Reply:
The indirection already exists. To know what these mean, I have to look at the documentation OR read the full context of the code which, in its terseness and unfamiliar language, adds cognitive load. With a named constant, you can convey what they mean in code and provide clues for readers of your code.

Reply:
I think we'll have to agree to disagree on this point. How does everything else look?